x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -20,19 +32,137 @@ interface ContentStreamDocument {
   if_seq_no?: number;
 }
 
+interface ChunkOutput {
+  chunk: number;
+  content: string;
+}
+
+interface ChunkSource {
+  parent_id: string;
+  output: ChunkOutput;
+}
+
 export class ContentStream extends Duplex {
-  private buffer = '';
+  /**
+   * @see https://en.wikipedia.org/wiki/Base64#Output_padding
+   */
+  private static getMaxBase64EncodedSize(max: number) {
+    return Math.floor(max / 4) * 3;
+  }
+
+  /**
+   * @note Raw data might be escaped during JSON serialization.
+   * In the worst-case, every character is escaped, so the max raw data length is twice less.
+   */
+  private static getMaxJsonEscapedSize(max: number) {
+    return Math.floor(max / 2);
+  }
+
+  private buffer = Buffer.from('');
+  private bytesRead = 0;
+  private chunksRead = 0;
+  private chunksWritten = 0;
+  private jobContentEncoding?: string;
+  private jobSize?: number;
+  private jobType?: string;
+  private maxChunkSize?: number;
+  private puid = new Puid();
   private primaryTerm?: number;
   private seqNo?: number;
 
-  constructor(private client: ElasticsearchClient, private document: ContentStreamDocument) {
+  /**
+   * The number of bytes written so far.
+   * Does not include data that is still queued for writing.
+   */
+  bytesWritten = 0;
+
+  constructor(
+    private client: ElasticsearchClient,
+    private exportTypesRegistry: ExportTypesRegistry,
+    private logger: LevelLogger,
+    private document: ContentStreamDocument
+  ) {
     super();
   }
 
-  async _read() {
+  private async getJobType() {
+    if (!this.jobType) {
+      const { id, index } = this.document;
+      const body: SearchRequest['body'] = {
+        _source: { includes: ['jobtype'] },
+        query: {
+          constant_score: {
+            filter: {
+              bool: {
+                must: [{ term: { _id: id } }],
+              },
+            },
+          },
+        },
+        size: 1,
+      };
+
+      const response = await this.client.search<ReportSource>({ body, index });
+      const hits = response?.body.hits?.hits?.[0];
+      this.jobType = hits?._source?.jobtype;
+    }
+
+    return this.jobType;
+  }
+
+  private async getJobContentEncoding() {
+    if (!this.jobContentEncoding) {
```

Comment:
we should make encoding an option and remove dependency from reporting. this way this stream will be reusable outside of reporting (and will be simpler).